Migrate to use v0.14 API

cosmo0920 opened the pull request:
I've tried to migrate to use v0.14 Output and Filter Plugin API.
This PR contains major update change and also includes in breaking changes.
If above questions/problems are acceptable or reasonable for you, could you bump up major version if releasing new version of gem?
And please refer http://docs.fluentd.org/v0.14/articles/plugin-update-from-v12 before release new version.

Thanks in advance.

cosmo0920 committed:
Migrate to use v0.14 API

cosmo0920 committed:
Remove legacy code

cosmo0920 committed:
Depends on FLuentd v0.14 or later

cosmo0920 committed:
Use v1 syntax in document

cosmo0920 committed:
Update supported Ruby versions

cosmo0920 committed:
Use event_time test helpers

cosmo0920 commented on lib/fluent/plugin/filter_eval.rb:
```diff
@@ -43,17 +45,13 @@ def configure(conf)
-        end
-      }
-      new_es
+    def filter(tag, time, record)
```
Note: In v0.14, `#filter` is faster than `#filter_stream` because v0.14 makes filter pipeline: https://github.com/fluent/fluentd/blob/26c4dc4872b71665cd743c1d6b71c774429c750f/lib/fluent/event_router.rb#L214-L232

ephemeralsnow replied:
How about replacing the body of the `filter` method with the body of the `filter_record` method?

It seems possible to delete redundant error handling.
https://github.com/fluent/fluentd/blob/161e1706273f2e9e179a0cc3d33e88a40dd60198/lib/fluent/plugin/filter.rb#L64-L66

cosmo0920 replied:
I've replaced it with the body of `#filter_record`.

ephemeralsnow commented:
I have not thought about the release of the new version yet, but before that, I'd like to think about migrating to a new configuration method.

`config_section :filter, multi: true`
like as
https://github.com/fluent/fluentd/blob/162031fadf6de19f8f7c0c40766f9c570f607baf/lib/fluent/plugin/filter_grep.rb#L34-L46

cosmo0920 committed:
Replace body with `#filter_record` method

cosmo0920 committed:
filter_eval: Use `<filter>` config section to specify filter rules

cosmo0920 committed:
out_eval_filter: Use `<filter>` config section to specify filter rules

cosmo0920 committed:
Modernize documentation

cosmo0920 committed:
Use array type instead of string with `#split`

Because string with `#split` makes the following a pitfall:

```aconf
requires yaml, time
<filter>
  filter "record.to_yaml; [Time.now.to_i, record]"
</filter>
```

is not valid config.
`requires yaml, time` will be interpreted as
`['yaml', ' time']` and execute:

```ruby
require 'yaml'
require ' time'
```

This is because why string type with `#split` makes a some pitfall.

cosmo0920 commented:
I've added commits which are reflected your review.
Please take a look.

ephemeralsnow commented:
Thanks for all your great work.
I'm almost satisfied, but I noticed that there are many `filter` keywords.
Please give me some time to think about keywords.


```
<filter **>
  @type eval
  <filter>
    filter "[time, record] if record['status'] == '404'"
  </filter>
  <filter>
    filter "[time, record] if record['status'] == 'POST'"
  </filter>
</filter>
```

cosmo0920 commented:
How about using `<rule>` instead of `<filter>` in specifying filter config section?

ephemeralsnow commented:
I think that's fine.
However, since `configN` and` filterN` are not pairs, I noticed that the current syntax is not appropriate.
I'm sorry but I will think about it for a while.

cosmo0920 commented:
Oh, I see....

Then, how about the following?

```aconf
<rule>
  filter [something filtering config]
</rule>
<eval>
  config [something evaluate instances]
</eval>
```

cosmo0920 committed:
Decomposite filter config section into rule and eval
